File: src/main/java/com/vc/medium/L938.scala
package com.vc.medium

class TreeNode938(var _value: Int) {
  var value: Int = _value
  var left: TreeNode938 = null
  var right: TreeNode938 = null
}

object L938 {
  def rangeSumBST(root: TreeNode938, L: Int, R: Int): Int = {
    var sum = 0
    def solve(root:TreeNode938): Unit = {
      import scala.collection.mutable
      val st = new mutable.Stack[TreeNode938]()
      var current = root
      while(current != null || st.nonEmpty) {
        if(current != null) {
          if(current.value >= L && current.value <= R) sum += current.value
          st.push(current)
          current = current.left
        }
        else current = st.pop.right
      }
    }
    solve(root)
    sum
  }
}

File: src/main/java/com/vc/easy/L938.java
package com.vc.easy;

import com.vc.hard.TreeNode;

class L938 {
    int res = 0;
    public int rangeSumBST(TreeNode root, int L, int R) {
        if(root == null) return 0;
        helper(root, L, R);
        return res;
    }

    private void helper(TreeNode root, int L, int R) {
        if(root == null) return;

        if(root.val < L) helper(root.right, L, R);
        else if(root.val > R) helper(root.left, L, R);
        else {
            res += root.val;
            helper(root.left, L, R);
            helper(root.right, L, R);
        }
    }
}
